### DISCUSSION TITLE
One-way collision detection strategy

### DISCUSSION THREAD

**Discussion by ewal** [excaliburjs/Excalibur#2577] [Q&A]
Hi, I'm trying to find a good strategy on how to resolve one-way collisions within a Tiled object layer.
I have a solution working but every now and then the collision check fails and the main actor is allowed to pass through the colliding object actor.
Each object has a `Fixed` `CollisionType` by default on which I toggle the `collisionType` depending on the colliding side.
The event listeners is hooked up in the `Scene`s onInitialize hook. Example below...

```js
// this === Scene
const oneWayActors = this.actors.filter(
  (actor) => actor.name === 'OneWay'
);

oneWayActors.forEach((actor) => {
  actor.on('precollision', (event) => {
    if (event.side === Side.Top) {
      event.target.body.collisionType = CollisionType.Fixed;
    } else {
      event.target.body.collisionType = CollisionType.Passive;
    }
  });
});
```

As stated, this works most of the times, but, if I stress the collider with enough events the actor will eventually fall through.
I'm trying to debug the problem but I'm rather suspecting that my strategy is faulty...

Any ideas? :)

**Comment by eonarheim** [comment]
@ewal I was thinking something similar as a solution! 

I bet you might have more robust results if you used raycasting from the actor to the platform, then deciding it's collision type (or perhaps collision group?). For example if the actor is above the platform, a downward ray could detect that and make the platform collision fixed, but if the platform is above the downward ray would not be intersecting the platform and could be passive?

The latest alpha has arbitrary raycasting against a scene built in that might make this a bit easier to code (non-alpha you'll need keep a reference to all the colliders you intend to raycast against) https://github.com/excaliburjs/Excalibur/pull/2410

I'll have some time later to try working out a sample 👍

**Reply by eonarheim** [reply]
Another idea might be to put an invisible collider above the platform to act like a sensor, if the player and the sensor are intersecting then the platform is solid?

**Reply by ewal** [reply]
Thank you for the reply. I will definitely look into raycasting and would very much appreciate an example. I've read about the concept but have real no experience working with it.

I actually use an invisible collider, about 1px in height.

**Reply by ewal** [reply]
I just did some more testing and noticed that the problem I experience also occurs on solid tile layers... I also noticed that I ran my logic in the `onPreUpdate` hook so I changed it to `onPostUpdate` but no change, maybe a bit worse actually. I'm reading about the hooks now but haven't really figured them out yet :)

**Reply by eonarheim** [reply]
@ewal I have something that works using a child actor as a sensor, it does take a little tuning depending on the size of your characters. I've highlighted the invisible sensor in purple.

![image](https://user-images.githubusercontent.com/612071/219269063-657c53de-4d75-4eef-8004-c321e7f25829.png)

We've been working on tuning up the platformer sample, I've added this example to that branch https://github.com/excaliburjs/sample-platformer/blob/presentation/src/one-way-platform.ts

I'll play around and see if I can get something with raycasting working soon.
 
![one-way-platform](https://user-images.githubusercontent.com/612071/219268175-8e781f6d-b599-4e98-adb3-c111c6f763f6.gif)

```typescript
import { Actor, ColliderComponent, CollisionType, Color, Engine, Entity, Shape, vec } from "excalibur";

export class OneWayPlatform extends Actor {
    private _sensor!: Entity;
    constructor(x: number, y: number, width: number, height: number) {
        super({
            x,
            y,
            width,
            height,
            color: Color.Red,
            collisionType: CollisionType.Passive
        });
    }

    onInitialize(_engine: Engine): void {
        this._sensor = new Actor({
            y: -20,
            collider: Shape.Box(this.width, 10, vec(0.5, 1))
        });

        const collider = this._sensor.get(ColliderComponent)!;
        this.addChild(this._sensor);
        collider.events.on('collisionstart', () => {
            this.body.collisionType = CollisionType.Fixed;
        });
        collider.events.on('collisionend', () => {
            this.body.collisionType = CollisionType.Passive;
        });
    }
}
```

**Reply by ewal** [reply]
Much appreciated! I will test working with a child actor.

I do however starting to suspect that the real problem is related to the movement controls. I will rethink them and perhaps create a new post about that subject.

**Comment by ewal** [comment]
@eonarheim, I'm starting to suspect that my problem is related to collision handling in the Tiled map. I stripped down a project to it's bare essentials in my hunt to locate the issue but I still experience it. The project is found here https://github.com/ewal/collision-demo if you would like to take a look. I added instructions in the Readme on how to recreate the collision issue.

**Reply by eonarheim** [reply]
@ewal thanks for the recording and sample, this is perfect! I have a couple theories about what is going but will confirm when I get a chance to dig in deeper. Excalibur moves actors in discrete steps which makes this type of collision problem tricky to solve completely. This gets extra tricky when your framerate is variable because the update distance becomes variable as well and can cause these intermittent issues.

Based on the videos in your sample here are my initial thoughts:

1. It might be that the falling velocity is high enough for the updates to step over or into the platform colliders entirely. Possible solutions (may want to use one of many of these): 
    *  Cap maximum velocity that your player can have in their post update, it will reduce the chance you update position through a platform (cheapest solution)
    * Set a fixed physics step, this will guarantee that the physics logic runs at a specific rate regardless of the graphics FPS during the simulation which will improve the collisions fidelity (at the cost of compute/fps). Some docs on this https://excaliburjs.com/docs/fixed-update/#implications    
      ```typescript
      var game = new ex.Engine({
        fixedUpdateFps: 60,
      })
      ```
     * Implement a type of lo-fi continuous collision by raycasting from the player in the direction of velocity, and if player would collide this frame based on the raycast, snap the player to the platform and zero out the vertical component of velocity.

**Reply by ewal** [reply]
Thank you for the reply. I had the `fixedUpdateFps` set to 60 and tried with capping the y-velocity to about half the gravity but I still experience random collision detection failures, even at lower velocities. Raycasting is next on my list but I have to do some personal leveling first :)

There are two other things in my approach that I'm a bit uncertain of is good practice when manipulating the actor.
Firstly I convert the delta so I can work with seconds inside the loop:
16.666666666666668 ms becomes  0.01666666666666667 seconds with the formula `delta * 0.001`.
The other thing I do is to manipulate both the actor's `velocity` _and_ `position`. My gut feeling tells me this is bad but I don't yet know why :)

I've also been tinkering with `velocityIterations` but without any real luck.

**Reply by eonarheim** [reply]
Apologies, that's definitely unexpected. I'll take a closer look at the collision logic when I get a chance. I've been a bit busier than usual so I'll try to dig in this week.

Tinkering `velocityIterations` will improve resolution reactions theoretically, it might help? `positionIterations` will improve overlap resolution quality (useful for object stacking).

It's possible that with the discrete nature of the physics simulation we'll need to do raycasting to counter this. I should be able to reproduce with your example and come up with some raycasting sample code.

**Reply by ewal** [reply]
No stress at all, just thankful for the help :raised_hands:

**Reply by ewal** [reply]
Noticed something interesting now. In certain x-coordinates the collision detection (right?) gets a weird flickering state. Could this explain the randomness that I'm experiencing?

[Screencast from 2023-03-03 09-07-55.webm](https://user-images.githubusercontent.com/676873/222666627-976ee513-6c33-4cf5-81e9-041c5fa6dd65.webm)

**Reply by eonarheim** [reply]
Hi @ewal 

I think I have 3 possible solutions for you! I've tested these on your example and they seem to work okay (with some fiddling).

1. Turn up the number of physic steps every frame (I found that I couldn't reproduce the issue at 80 fps physics) the downside to this is that you'll be running more physics simulations every frame, but the collision fidelity should be much better.
    ```typescript
    class Game extends Engine {
      constructor() {
        super({
          displayMode: DisplayMode.FitScreen,
          width: 40 * GameConfig.tileSize,
          height: 30 * GameConfig.tileSize,
          antialiasing: false,
          snapToPixel: false,
          canvasElementId: 'game',
          suppressConsoleBootMessage: true,
          suppressPlayButton: true,
          fixedUpdateFps: 80, // 80 fps physics 
        });
      }
      ...
    }
    ```

2.  Clamp the player velocity to some maximum value, this seems to work pretty well? I'm actually thinking of making this type of thing a built feature of Excalibur, an optional speed limit on Actors. This needs some tuning to get right, but in this case it seems that a max of 450 pixels/sec made this issue hard to reproduce.

    ```typescript
      override onPreUpdate(_engine: Engine, delta: number): void {
        ...
        // Clamp the magnitude of the velocity to prevent clipping
        this.vel.size = clamp(this.vel.size, 0, 450);
      }
    ```

3. Continuous collision with raycasting! Excalibur should probably have a setting to do this, but this seems like the most robust solution. This takes advantage of the latest alpha feature of arbitrary scene raycasting, so not available in v0.27 unfortunately. The gist is when the player starts moving fast enough to start causing problems for the discrete physics system we start raycasting into the scene to detect a point of impact, then move the player there.

    ```typescript
      private snapPlayerPositionByRayCast(seconds: number) {
        // Pixel delta that we'll move this frame plus some extra
        const nextFramePositionDelta = this.vel.scale(seconds * 1.5).add(Physics.acc.scale(.5 * seconds * seconds));
    
        // this._dir = nextFramePositionDelta;
    
        // If we'll move more than half the hight of the collider we are at risk of clipping
        if (nextFramePositionDelta.size >= this.height / 2) {
    
          const dir = this.vel.normalize();
          const ray = new Ray(this.pos, dir);
    
          // Raycast to find any walls that we'll clip
          const hits = this.scene.physics.rayCast(ray, {
            maxDistance: nextFramePositionDelta.size,
            searchAllColliders: true
          });
    
          // Since we are casting from the center this will hit the player collider
          const otherColliders = hits.filter(hit => hit.collider != this.collider.get());
    
          // Any hit?
          if (otherColliders.length) {
            console.log('acc', this.acc);
            console.log('vel', this.vel);
            console.log('pos', this.pos);
            console.log('time', nextFramePositionDelta.size);
            console.log('hit', otherColliders[0].point);
    
            // snap the actor to the surface platform
            this.pos.y = otherColliders[0].point.y - this.height / 2;
            // negate acceleration
            this.vel.y = -Physics.acc.scale(.5 * seconds * seconds).y;
            return true;
          }
        }
        return false;
      }
    
      override onPreUpdate(_engine: Engine, delta: number): void {
        ...
        // if we don't snap player do the normal position update
        if (!this.snapPlayerPositionByRayCast(deltaTime)) {
          // Position
          this.pos.x += this.vel.x * deltaTime;
          this.pos.y += this.vel.y * deltaTime;
        }
    
        // "Jumping" (not really. Just moving up for demo purpose)
        if (InputState.getInstance().isHoldingJump) {
          this.vel.y = -(250 + Math.abs(this.vel.x) * 0.5);
        }
      }
    
    ``` 

Let me know if this helps, I can push up a fork of your repo too if that helps 👍

**Reply by ewal** [reply]
Just wow! Thank you so much @eonarheim :raised_hands: I will try the raycast example as soon as I can make time.
